src/ai/backend/agent/docker/kernel.py

```diff
@@ -127,6 +132,51 @@ async def get_service_apps(self):
         result = await self.runner.feed_service_apps()
         return result
 
+    def _get_lock_path(self, commit_path: Path):
+        return commit_path / "lock"
+
+    async def check_duplicate_commit(self, path: Path):
+        lock_path = self._get_lock_path(path)
+        if lock_path.exists():
+            return CommitStatus.DUPLICATED
+        return CommitStatus.AVAILABLE
+
+    async def commit(self, path: Path):
+        assert self.runner is not None
+
+        container_id: str = str(self.data["container_id"])
+        now = datetime.now(tzutc()).strftime("%Y-%m-%dT%H:%M:%S")
+        filename = f"{now}.tar.gz"
+        filepath = path / filename
+        try:
+            Path(filepath).parent.mkdir(exist_ok=True, parents=True)
+        except ValueError:  # parent_path does not start with work_dir!
+            raise AssertionError("malformed committed path.")
+        loop = current_loop()
+        lock_path = self._get_lock_path(path)
+        try:
+            async with FileLock(path=lock_path, timeout=0.1):
+                docker = Docker()
+                log.info("Container is being committed to {}", filepath)
+                async with docker._query(
+                    f"containers/{container_id}/export",
+                    method="GET",
+                ) as response:
+                    log.debug("Container commit data is being written on {}", filepath)
+                    data: bytes = await response.read()
+
+                    def _save_tar():
+                        tar_info = tarfile.TarInfo(filename)
+                        tar_info.size = len(data)
+                        with tarfile.open(name=filepath, mode="x:gz") as tar:
+                            tar.addfile(tar_info, fileobj=BytesIO(data))
```

Comment:
This will load all the output in the memory and write it once, and is highly likely to cause out-of-memory of the host system.
Make it read a chunk and write a chunk. Refer the sample codes (this code is the server-side upload receiver but you can implement chunking in the same way in the client-side download receiver):
https://github.com/lablup/backend.ai/blob/141844d383970b7274b9861b625d7f51a6601998/src/ai/backend/storage/api/client.py#L245-L266